From 44815bd61a5fcb241812c4bf033f56d226ec1e18 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Thu, 27 Jul 2017 00:08:15 +0800
Subject: [PATCH 2290/5242] MLK-16013-11 staging: typec: only clear alert
 events for enabled irq

commit  74b58131ad36025d5f2032d59725d2b033321a28 from
https://source.codeaurora.org/external/imx/linux-imx.git

Uncondtionaly clear all events may impact the typec controller state
machine, so save the enabled irq and only clear alert events for enabled
irq.

Reviewed-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/staging/typec/tcpci.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/typec/tcpci.c b/drivers/staging/typec/tcpci.c
index 57a70fc..76492c2 100644
--- a/drivers/staging/typec/tcpci.c
+++ b/drivers/staging/typec/tcpci.c
@@ -30,6 +30,7 @@ struct tcpci {
 
 	struct tcpc_dev tcpc;
 	struct tcpci_data *data;
+	unsigned int irq_mask;
 };
 
 struct tcpci_chip {
@@ -397,6 +398,8 @@ static int tcpci_init(struct tcpc_dev *tcpc)
 		TCPC_ALERT_RX_HARD_RST | TCPC_ALERT_CC_STATUS;
 	if (tcpci->controls_vbus)
 		reg |= TCPC_ALERT_POWER_STATUS;
+	tcpci->irq_mask = reg;
+
 	return tcpci_write16(tcpci, TCPC_ALERT_MASK, reg);
 }
 
@@ -407,10 +410,10 @@ irqreturn_t tcpci_irq(struct tcpci *tcpci)
 	tcpci_read16(tcpci, TCPC_ALERT, &status);
 
 	/*
-	 * Clear alert status for everything except RX_STATUS, which shouldn't
+	 * Clear alert status for enabled irq except RX_STATUS, which shouldn't
 	 * be cleared until we have successfully retrieved message.
 	 */
-	if (status & ~TCPC_ALERT_RX_STATUS)
+	if ((status & ~TCPC_ALERT_RX_STATUS) & tcpci->irq_mask)
 		tcpci_write16(tcpci, TCPC_ALERT,
 			      status & ~TCPC_ALERT_RX_STATUS);
 
-- 
1.7.9.5

